blob/store.go

```diff
@@ -24,11 +24,28 @@ import (
 	"sort"
 )
 
-// A KV represents a mutable key-value store in which each value is identified
-// by a unique, opaque string key.  An implementation of KV is permitted (but
-// not required) to report an error from Put when given an empty key.  If the
-// implementation cannot store empty keys, it must report ErrKeyNotFound when
-// operating on an empty key.
+// A Store represents a collection of key-value namespaces ("keyspaces")
+// identified by string labels. Each keyspace in a store is logically distinct;
+// the keys from one space are independent of the keys in another.
+//
+// Implementations of this interface must be safe for concurrent use by
+// multiple goroutines.
+type Store interface {
+	// Keyspace returns a key space on the store.
+	Keyspace(name string) (KV, error)
+
+	// Close allows the store to release any resources held open while in use.
+	// If an implementation has nothing to release, it must report nil.
+	// After closing a Store, any keyspaces derived from the store should report
+	// errors when used.
+	Close(context.Context) error
+}
+
+// A KV represents a mutable set of key-value pairs in which each value is
+// identified by a unique, opaque string key.  An implementation of KV is
+// permitted (but not required) to report an error from Put when given an empty
+// key.  If the implementation cannot store empty keys, it must report
+// ErrKeyNotFound when operating on an empty key.
```

Comment:
API question: should KVs have an API to create sub-KVs under them?

What I'm picturing is a hypothetical file API shaped differently to the existing one: a `Filesystems` type contains a consistent set of both CASed file trees and associated root pointers (this object is where GC() would be implemented, for example). It's constructed by handing it a single KV, which the `Filesystem` then further partitions into two KVs, one for the CAS trees and one for the root pointers.

Arguably this is bad because it creates a parallel hierarchical concept other than files... But for API encapsulation, it seems very handy to be able to take in one "here's a place to plop your blobs" and break into several places to plop one's blobs.

Reply:
Updated to add a `Sub` method, PTAL! 🙂